### PR TITLE
GNN Network for Multi-Class Node Classification on Facebook Page Data

### PR DESCRIPTION
**Pull request by TrazZed**
# Multi-layer graph neural network (GNN) for semi supervised multi-class node classification using Facebook Large Page-Page Network Data
This Pull Request adds an implementation of a Graph Convolutional Network (GCN) to perform multi-class node classification on the Facebook Large Page-Page Network Data. Details regarding implementation and results are presented in the README.md

# Files
- ```README.md```: This provides documentation about the problem, including theory into Graph Neural Networks, Model Archiecture for the problem, and results
- ```assets/```: Images that were acquired from referenced sources and used in the README.md for support
- ```outputs/```: Images that were produced from the associated code and used in the README.md for results and analysis
- ```dataset.py```: Implementation of class for processing data from .npz file
- ```modules.py```: Implementation of the GCN used for the model
- ```train.py```: Implementation of the training and validation loop for training and saving the model
- ```predict.py```: Implementation of taking a saved model and using it to classify nodes, as well as create plots used for results analysis
- ```utils.py```: Helper functions used for plotting results and saving images

Author: Tristan Hayes (46969842).

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
GNN FB Page-Page Network dataset ---- Normal Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 4.5 | Some overfitting seen. No early stop|
| Implementation functions as intended	| 3 | 3 | 3 layer GCN, torch_geometric/GCNConv |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 0 | Normal Difficulty |
| Section IV : Max mark 15 from 20		|   | **14.5**   | |

**TSNE / UMAP:** 
- Ok, though would have expected to see a bit more separation at the reported accuracy.
- Discussion: Good, with a bit more depth useful.

**Suggestions:**
Not mentioned the train acc accuracy, this is a useful metric to know the generalization of the model.
Could add a hyper parameter search or discuss how the values were obtained.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Some overfitting seen. No early stop</td><td>-2</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. Give a  high level intro to the approach instead of the file list.</td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-7</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
